### DISCUSSION TITLE
Removing a string from a list in the nf script section with groovy - multiple solutions do not work

### DISCUSSION THREAD

**Discussion by colindaven** [nextflow-io/nextflow#3891] [Q&A]
Hi,
I have problems with groovy code apparently not working in the nextflow script section. I'm just trying to remove a string from a groovy list by multiple methods, either by removing it directly, or adding a string at position 1 and then eg calling `.unique()` etc.

Why does arg.unique() not work in Nextflow on a list of strings??  Could it be that the filenames are not actually just ArrayLists of strings in Nextflow ? 

**Edit  and resolution** - yes, this was actually the case - check the class of your object before wondering why groovy standard list methods are not working!

Example -  I expected the nextflow object `nf_fasta_paths` to be a list, but it was not. All groovy methods such as `contains(), unique() `etc failed without a warning/error in this code. Everything works as expected if I convert the object `nf_fasta_paths` into a standard groovy `ArrayList` in a for loop.

```groovy
    input:
    path nf_fasta_paths

    output:
    path "*.rgfa", emit: rgfa
      
    script:
    prefix = params.reference_fasta.toString().tokenize('.').get(0)
    name = params.reference_fasta
    ref_name = params.reference_fasta.toString()
    println("ref_name: " + ref_name)

    
    println ref_name.getClass()     //     class java.lang.String
    println nf_fasta_paths.getClass()   //     class nextflow.util.BlankSeparatedList

    // The nf paths are a nextflow.util.BlankSeparatedList, not a java or groovy standard arraylist. This causes havoc - so convert
    fasta_list = []
    for (str in nf_fasta_paths) {
        fasta_list.add(str)
    }
    
    // unique works now
    fasta_list.unique()



```

**Comment by mribeirodantas** [comment]
Good to know you solved your issue. A snippet I like to share when we want to know what methods are supported by an object in a Nextflow channel can be found below:

```Groovy
some_ch = Channel.of(file("/Users/mribeirodantas/dag.dot"), "some random string")
some_ch.view { it.properties.each{ println it } }
```

**Comment by bentsherman** [comment]
I think the issue is that a path input can be a file or a list of files depending on whether there are multiple files, which has caused lots of headaches for users over the years. This PR #3706 will address the problem by allowing you to specify the number of files you expect (e.g. `1..*`) and that way you will know for certain whether it is a list or a single file.

**Reply by colindaven** [reply]
As a user you're not sure where nextflow ends and groovy/java methods start, especially if you haven't done much object oriented programming before (I've done a bit). So maybe a doc on finding out what methods are available for each object could be added to the nextflow scripting docs section. Having said that, the docs are already really, really great.

Thanks for your comments.

**Reply by bentsherman** [reply]
No you're right, I share that frustration. Now that I've been working on Nextflow for a while, I have learned those boundaries, but I want to create some kind of reference docs for what is available from Groovy versus what is added by Nextflow, without e.g. recreating the entire Groovy docs.
